Refactor Git metadata collection

dustinbyrne opened the pull request:
Git metadata collection is refactored in order to:
- always return some sort of data, even if an error occurs
- remove fallback strategy attempting to resolve the last parent of a merge commit to a branch name (use the `--branch` or `-b` argument instead)

Also:
- clean up the error message when no branch flag is provided or resolvable from Git

dividedmind commented on cmd/upload.go:
```diff
@@ -80,6 +80,11 @@ func init() {
 					progressBar.Add(1)
 				}
 
+				if branch == "" && git.Branch == "" {
```
Should we _require_ branch information? That doesn't seem very friendly, especially without a default.

dustinbyrne replied:
It's now optional as long as there's no associated commit

kgilpin commented:
When uploading AppMaps from a directory that does not contain Git metadata, the following error occurs:

```
➜  tmp ~/source/appland/appland-cli/appland-cli upload ./rspec/DAO_User_authentication_is_accepted_with_a_valid_password.appmap.json
warn: Could not collect git metadata (repository does not exist)
  50% |████████████████████                    |  [0s:0s]panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1541a18]

goroutine 1 [running]:
github.com/applandinc/appland-cli/cmd.init.3.func1(0xc000152840, 0xc000126a70, 0x1, 0x1)
	/Users/kgilpin/source/appland/appland-cli/cmd/upload.go:83 +0x548
github.com/spf13/cobra.(*Command).execute(0xc000152840, 0xc000126a50, 0x1, 0x1, 0xc000152840, 0xc000126a50)
	/Users/kgilpin/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x2aa
github.com/spf13/cobra.(*Command).ExecuteC(0x1b4e460, 0x0, 0x0, 0x0)
	/Users/kgilpin/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/kgilpin/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
github.com/applandinc/appland-cli/cmd.Execute()
	/Users/kgilpin/source/appland/appland-cli/cmd/root.go:51 +0x2d
main.main()
	/Users/kgilpin/source/appland/appland-cli/main.go:8 +0x20
```

So, I'd suggest that Git metadata be optional. And if there is no Git metadata, the Git options (url, branch, commit) are blank. The server will accept scenarios without Git metadata. The `-branch` option can be used to specify the branch, but it should not be an error if there is no Git metadata and no `-b` option is provided.

dustinbyrne commented:
I've changed it so that
  a.) we won't attempt to patch Git metadata into scenarios if the data is unavailable
  b.) both `commit` and `branch` must both be either present or omitted. Presence of only one will cause an error.

I've also added the ability to override the app name via an `--app`/`-a` flag.

kgilpin commented on cmd/upload.go:
```diff
@@ -80,9 +80,17 @@ func init() {
+				// either both commit and branch are specified or both are unspecified
+				// fail otherwise
+				commitProvided := bool(git != nil && git.Commit != "")
+				branchProvided := bool(git != nil && git.Branch != "" || branch != "")
```
I think operator precedence may bite you here, with the && higher than the ||

dustinbyrne replied:
I've rewritten it as `(git != nil && git.Branch != "") || branch != ""` for clarity. Functionally, it should be the same.
